### PR TITLE
Insert

### PR DESCRIPTION
**Pull request by RobStallion**
#45 

"alog-ifies" insert functionality, adding a CID made from the params and a unique entry_id from a substring of the CID. 

Adds tests for insert function

This step, as it is currently, allows for duplicate data to be inserted. Comments have been added to code for my thoughts on this matter.

### CODE REVIEW

**Comment by RobStallion** [comment]
@Danwhy I do not think that this PR is complete (at all) but I think that it might be in a place that it can be merged into the adapter branch. 

Please let me know if you have any thoughts on this.

**Review by nelsonic — COMMENTED** [review]
@RobStallion looking good so far. 👀 👍

**Comment by Danwhy** [comment]
I think if we add some tests, we may get a better idea of how we want it to actually work/what makes sense with regards to duplicate data etc.

You'll need to delete or comment out all the old tests that have `use Alog` in them to run any tests you write. Take a look at the update test on my `update` branch if you need any help getting started.

**Comment by RobStallion** [comment]
Testing the insert functions has 'brought to light' some things that we could (and maybe should) change in the future. Please see [this comment](https://github.com/dwyl/alog/issues/45#issuecomment-469233754) for more info on this.

**Review by Danwhy — APPROVED** [review]
Looks great. Definitely a good idea to get this in to the adapter branch :+1:

File: lib/alog.ex [hunk 1]
```diff
@@ -3,16 +3,72 @@ defmodule Alog do
...
+      |> Map.put(:entry_id, entry_id)
+      |> Enum.into([])
+
+    insert_logic(adapter_meta, source, prefix, params, on_conflict, returning, opts)
+  end
+
+  defp create_entry_id(source, adapter_meta, cid, n) do
+    entry_id = String.slice(cid, 0..n)
+    entry_id_query = "SELECT * FROM #{source} where entry_id='#{entry_id}'"
+    {:ok, results} = Ecto.Adapters.SQL.query(adapter_meta, entry_id_query, [])
```
Comment (SimonLab) [inline lib/alog.ex:52]: We might be able to enhance this existing cid lookup later on. Maybe we could use a cache system to store all the current cid (with ETS)? If we can manage to not send another query to Postgres that would be ideal but it's working for now :+1:

File: lib/alog.ex [hunk 2]
```diff
@@ -3,16 +3,72 @@ defmodule Alog do
...
+    params
+    |> Enum.into(%{})
+    |> Map.put_new(:inserted_at, NaiveDateTime.utc_now())
+    |> Map.put_new(:updated_at, NaiveDateTime.utc_now())
+  end
+
+  defp insert_logic(adapter_meta, source, prefix, params, on_conflict, returning, opts) do
+    {kind, conflict_params, _} = on_conflict
+    {fields, values} = :lists.unzip(params)
+    sql = @conn.insert(prefix, source, fields, [fields], on_conflict, returning)
```
Comment (SimonLab) [inline lib/alog.ex]: I'm not sure how/where the `@conn` module attribute is defined, I can't see it being done in the Alog module.
I think a `@conn` is defined before the adapter is being called but I think module attributes are scoped to the module and can't be use in other modules (didn't find any explanation on the scope of the module on the Elixir doc https://elixir-lang.org/getting-started/module-attributes.html)

@RobStallion do you know where it is defined?

**Review by SimonLab — COMMENTED** [review]
@RobStallion looks good, just a question about the module attribute `@conn`

File: lib/alog.ex (hunk 2 above)
Comment (RobStallion) [reply lib/alog.ex]: @SimonLab That's a really good spot. I meant to change that. It is defined [here](https://github.com/elixir-ecto/ecto_sql/blob/master/lib/ecto/adapters/sql.ex#L80) in the Ecto.Adapter.SQL module. As it is a defined in a macro it looks like the module __using__ it inherits the module attributes as well.

I agree that it is SUPER unclear though so I'll update it now 👍

Comment (SimonLab) [reply lib/alog.ex]: ok makes sense, so to recap because `Ecto.Adapters.SQL` module is `used` in `Alog`, all the module attributes in the macro are accessible on Alog.

```
defmodule Alog do
  use Ecto.Adapters.SQL,
```
and
```
  defmacro __using__(opts) do
    quote do
      @behaviour Ecto.Adapter
      @behaviour Ecto.Adapter.Migration
      @behaviour Ecto.Adapter.Queryable
      @behaviour Ecto.Adapter.Schema
      @behaviour Ecto.Adapter.Transaction

      opts = unquote(opts)
      @conn __MODULE__.Connection
...
```

**Review by SimonLab — APPROVED** [review]
After creating the table and running the migrations the tests are green :tada: 
image

Thanks @RobStallion  :+1:

File: lib/alog.ex [hunk 3]
```diff
@@ -3,16 +3,72 @@ defmodule Alog do
...
 
   @impl true
-  defdelegate storage_down(opts), to: Ecto.Adapters.Postgres
+  defdelegate storage_up(opts), to: EAP
+
+  @impl true
+  defdelegate storage_down(opts), to: EAP
+
+  # overrides insert/6 defined in Ecto.Adapters.SQL
+  def insert(adapter_meta, %{source: "schema_migrations", prefix: prefix}, params, on_conflict, returning, opts) do
```
Comment (SimonLab) [inline lib/alog.ex]: Just noticed that the `insert` function doesn't have the `@impl true`, do we need to add it as the adapter behaviour will need it?

File: test/alog_test.exs [hunk 4]
```diff
@@ -1,74 +1,74 @@
...
-  describe "Not compatible with unique index" do
-    test "Throws error if unique index exists" do
-      assert_raise RuntimeError, fn ->
-        %Alog.TestApp.Unique{}
-        |> Alog.TestApp.Unique.changeset(%{name: "unique item"})
-        |> Alog.TestApp.Unique.insert()
-      end
-    end
-  end
+  # use Alog.TestApp.DataCase
```
Comment (nelsonic) [inline test/alog_test.exs:2]: when commenting out an entire file, consider adding a line at the top explaining _why_.
otherwise the _next_ person reading it will scratch their heads and have to do _forensics_ to understand it. 🤔

**Review by nelsonic — APPROVED** [review]
@RobStallion insert is looking good. ✅ 
Please consider leaving a comment (_preferably linking to an issue/discussion_) explaining why several tests are commented out. 💬 
Happy to merge into adapter branch to continue work. 👍
